### ISSUE TITLE
Axios will not timeout on idle sockets

### ISSUE THREAD

**Issue by justinmeiners** [axios/axios#2143]
**Describe the bug**

On nodejs the timeout mechanism does not work correctly.
It is only concerned with the time to receive the first byte.
If the server sends a few bytes across the socket, but leaves it open
the request will never timeout.

This only happens with node, not in the browser.

**To Reproduce**

1. Create a web server which writes a few bytes and then stops :

```
const http = require('http');

const hostname = '127.0.0.1';
const port = 3000;

const server = http.createServer((req, res) => {
    res.statusCode = 200;
    res.setHeader('Content-Type', 'text/plain');

    res.write('h');
//    res.end();
});

server.listen(port, hostname, () => {
    console.log(`Server running at http://${hostname}:${port}/`);
});
```
2. Make a get request to the server with axios, with a timeout setting.

```
const axios = require('axios');

axios.get('http://localhost:3000', { timeout: 1000 })
.then(() => console.log('hello there'));
```

3. The request will never timeout.

**Expected behavior**
If a socket stops writing for long period of time, the request should timeout.

nodejs get requests can detect this. You need to use their `timeout`
event instead of just using a timer. 

https://nodejs.org/api/http.html#http_event_timeout

**Environment:**
 - Axios Version [e.g. 0.18.0]
 - OS: Linux (Ubuntu 18)

Looking at the code, I believe this is independent of environment.

**Comment by aarcangeli** [comment]
I agree, an infinite fetching time can be problematic, especially if you expect a small response.

My workaround is to use a cancel token.

    const CancelToken = axios.CancelToken;
    const source = CancelToken.source();
    let timeout = setTimeout(() => source.cancel('Timeout'), 5000);

    axios.get('http://localhost:3000', { timeout: 1000, cancelToken: source.token })
        .then(() => { clearTimeout(timeout); console.log('hello there') });

**Comment by justinmeiners** [comment]
Yes, I workaround it by just using the `http` module in node. Since it has support for proper timeout, I imagine adding it to axios would be easy. I feel this is especially important because it can have drastic consequences for the client.

**Comment by ryouaki** [comment]
maybe it looks like [#1752](https://github.com/axios/axios/pull/1752)

**Comment by Alanscut** [comment]
this issue has been fixed in v0.19.0, please try it.

**Comment by justinmeiners** [comment]
@Alanscut I am not longer using the library. If it is fixed I will close the issue.

**Comment by joquijada** [comment]
> I agree, an infinite fetching time can be problematic, especially if you expect a small response.
> 
> My workaround is to use a cancel token.
> 
> ```
> const CancelToken = axios.CancelToken;
> const source = CancelToken.source();
> let timeout = setTimeout(() => source.cancel('Timeout'), 5000);
> 
> axios.get('http://localhost:3000', { timeout: 1000, cancelToken: source.token })
>     .then(() => { clearTimeout(timeout); console.log('hello there') });
> ```

This workaround worked for me, I'm on Axios 0.21.1. This has wreaked havoc in my production app. We were on Axios 0.19.x without any issues. It took me 3 days to isolate where the file descriptor and memory leak were coming from

**Comment by justinmeiners** [comment]
@joquijada your fix requires the entire request to occur within that time. I believe the nice thing about node is it considers the time between bytes.

**Comment by jasonsaayman** [comment]
This has been fixed with #3694

**Comment by jesse23** [comment]
+2 cents to this issue:
```
Test env: macOS Monterey
Node version: v16.13.1

0.18.0 -> not working for test case at the top even with workaround.
0.21.4 -> woks fine even without workaround
```
